internal/server/logbuffer/merge.go

```diff
@@ -0,0 +1,146 @@
+package logbuffer
+
+import (
+	"io"
+	"time"
+)
+
+// Merger can combine multiple log streams into one stream. It presumes
+// each stream emits entries in time order, and then weaves the entries
+// together to create a total time ordered stream.
+type Merger struct {
+	readers []MergeReader
+	heads   []TimedEntry
+}
+
+// TimedEntry is the interface each input returns entries in
+type TimedEntry interface {
+	// Time is the associated time of the value. This is used to sort
+	// the entry against other entries from other inputs.
+	Time() time.Time
+
+	// Value returns the log entry itself. This allows time addition
+	// wrappers to return their original value.
+	Value() interface{}
+}
+
+// MergeReader is value that returns TimedEntry's for Merger
+// to weave together.
+type MergeReader interface {
+	Next() (TimedEntry, error)
+}
+
+// NewMerger creates a new Merger, with the stream generated
+// from the given inputs.
+func NewMerger(readers ...MergeReader) *Merger {
+	var lm Merger
+	lm.readers = readers
+
+	lm.heads = lm.makeHeads()
+
+	return &lm
+}
+
+// TimedEntries is a convience type of TimedEntry's that provides
+// the LogMergeInput interface.
+type TimedEntries []TimedEntry
+
+// Next returns the next value in the slice and then shrinks itself.
+func (t *TimedEntries) Next() (TimedEntry, error) {
+	if len(*t) == 0 {
+		return nil, io.EOF
+	}
+
+	ent := (*t)[0]
+
+	*t = (*t)[1:]
+
+	return ent, nil
+}
+
+// Create a slice to be used by refillEntries and findNext
+func (l *Merger) makeHeads() []TimedEntry {
+	return make([]TimedEntry, len(l.readers))
+}
+
+// Populate any missing entries with an entry from the corresponding
+// input (index in slice corresponds to inputs slice).
+func (l *Merger) refillEntries(entries []TimedEntry) int {
+	var pop int
+
+	for i, ent := range entries {
+		if ent != nil {
+			pop++
+			continue
+		}
+
+		ent, err := l.readers[i].Next()
+		if err == nil {
+			pop++
+			entries[i] = ent
+		}
+	}
+
+	return pop
+}
+
+// Find the entry in entries with earliest time, returning the entry and the
+// input that generated it.
+func (l *Merger) findNext(entries []TimedEntry) (TimedEntry, MergeReader) {
+	var (
+		best      TimedEntry
+		bestIdx   int
```

Comment:
nit: maybe `earliest` and `earliestIndex` would be clearer?

Reply:
I feel like best is clearer.